deeppavlov/pipeline_manager/observer.py

```diff
@@ -0,0 +1,180 @@
+
+        # tmp parameters
+        self.pipe_ind = 0
+        self.pipe_conf = None
+        self.model = None
+        self.pipe_res = None
+        self.pipe_time = None
+        self.batch_size = None
+
+        # build folder dependencies
+        self.save_path = self.root / self.date / self.exp_name / 'checkpoints'
+        self.log_path = self.root / date / self.exp_name
+        self.exp_file = self.log_path / (self.exp_name + '.json')
+        self.log_file = self.log_path / 'logs.jsonl'
+
+        if not self.save_path.is_dir():
+            self.save_path.mkdir(parents=True)
+        if self.plot:
+            if not (self.log_path / 'images').is_dir():
+                (self.log_path / 'images').mkdir()
+
+        self.exp_info = OrderedDict(date=date,
+                                    exp_name=self.exp_name,
+                                    root=str(self.root),
+                                    info=self.exp_inf,
+                                    number_of_pipes=None,
+                                    metrics=None,
+                                    target_metric=None,
+                                    dataset_name=None)
+        self.log = None
+
+    def save_exp_info(self):
+        """ Write exp_info in json file. """
+        with open(str(self.exp_file), 'w') as exp_file:
+            json.dump(self.exp_info, exp_file)
+
+    def exp_time(self, time: str) -> None:
+        """
+        Adding the time duration of the experiment in log file.
+
+        Args:
+            time: the time duration of the experiment
+
+        Returns:
+            None
+        """
+        with open(str(self.exp_file), 'r') as old_log:
+            old_log = json.load(old_log)
+
+        old_log['full_time'] = time
+        with open(str(self.exp_file), 'w') as log_file:
+            json.dump(old_log, log_file)
+
+    def tmp_reset(self) -> None:
+        """ Reinitialize temporary attributes. """
+        # tmp parameters
+        self.pipe_ind = 0
+        self.pipe_conf = None
+        self.model = None
+        self.pipe_res = None
+        self.pipe_time = None
+        self.batch_size = None
+        self.log = None
+
+    def write(self) -> None:
+        """ Write pipeline logs in jsonl file. """
+        if self.log_file.is_file():
+            with open(str(self.log_file), 'a') as logs:
+                logs.write(json.dumps(self.log))
+                logs.write('\n')
+        else:
+            with open(str(self.log_file), 'w') as logs:
+                logs.write(json.dumps(self.log))
+                logs.write('\n')
+
+    def update_log(self):
+        """ Updates a log with new pipeline information. """
+
+        if (self.model is None) and (self.pipe_conf is not None):
+            for component in self.pipe_conf:
+                if component.get('main') is True:
+                    self.model = component['component_name']
+
+        pipe_name = '-->'.join([x['component_name'] for x in self.pipe_conf])
```

Comment:
There is a check above that implies that `self.pipe_conf` can be `None`, but here it's just ignored.

Reply:
Yes, thanks. All condition was unnecessary. Apparently I forgot about it.